### PR TITLE
add collector support activemq

### PR DESCRIPTION
**Pull request by tian-junwei**
like collector-rabbitmq the collector-activemq support collector consumes a ActiveMQ queue for messages that contain a list of spans

### CODE REVIEW

**Comment by codefromthecrypt** [comment]
@IAMTJW I like your energy and eagerness to do work. We have a change policy though which suggests multiple people need to buy-in before we make a standard feature. For example, adding this here also has impact as people using tracers will also want it. there's a lot of work around it.

I see that there are a couple votes here which is good. https://github.com/openzipkin/zipkin/issues/1990

We do need buy-in from folks in @openzipkin/core before adding this. Let's try to see if anyone is interested. Whatever we add becomes years of support work, so it is only fair to make sure people are ok with that.

**Comment by tian-junwei** [comment]
> @IAMTJW I like your energy and eagerness to do work. We have a change policy though which suggests multiple people need to buy-in before we make a standard feature. For example, adding this here also has impact as people using tracers will also want it. there's a lot of work around it.
> 
> I see that there are a couple votes here which is good. #1990
> 
> We do need buy-in from folks in @openzipkin/core before adding this. Let's try to see if anyone is interested. Whatever we add becomes years of support work, so it is only fair to make sure people are ok with that.

get it , our company use activemq as collector，use postgresql as storage and I want to make it available to others.

**Comment by codefromthecrypt** [comment]
@IAMTJW would you be willing to contribute senders in zipkin-reporter-java and other languages for this?

File: zipkin-autoconfigure/collector-activemq/src/main/resources/zipkin-server-shared.yml [hunk 1]
```diff
@@ -0,0 +1,221 @@
+zipkin:
```
Comment (codefromthecrypt) [inline zipkin-autoconfigure/collector-activemq/src/main/resources/zipkin-server-shared.yml]: you don't need this file, I think you want to make activemq.yml right?

File: zipkin-collector/activemq/README.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+Property | Environment Variable | Description
+--- | --- | ---
+`zipkin.collector.activemq.connection-timeout` | `ACTIVE_CONNECTION_TIMEOUT` | Milliseconds to wait establishing a connection. Defaults to `60000` (1 minute)
+`zipkin.collector.activemq.queue` | `ACTIVE_QUEUE` | Queue from which to collect span messages. Defaults to `zipkin`
+
+If the URI is set, the following properties will be ignored.
+
+Property | Environment Variable | Description
+--- | --- | ---
+`zipkin.collector.activemq.addresses` | `ACTIVE_ADDRESSES` | Comma-separated list of ActiveMQ addresses, ex. `tcp://localhost:61616,tcp://localhost2:61616`
```
Comment (codefromthecrypt) [inline zipkin-collector/activemq/README.md]: is the naming convention `ACTIVE` normal in activemq or is it `ACTIVEMQ` or different? can you give example of some commandline tool which uses this convention?

File: zipkin-collector/activemq/src/main/java/zipkin2/collector/activemq/ActiveMQCollector.java [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+        builder.connectionFactory.setBrokerURL("failover:("+builder.addresses+")?initialReconnectDelay=100");
+        connection = builder.connectionFactory.createConnection();
+        Session session=connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+        connection.start();
+        Destination destination = session.createQueue(builder.queue);
+        MessageConsumer messageConsumer = session.createConsumer(destination);
+        Collector collector = builder.delegate.build();
+        CollectorMetrics metrics = builder.metrics;
+        messageConsumer.setMessageListener(new ActiveMQSpanConsumerMessageListener(collector,metrics));
+      }catch (Exception e){
```
Comment (codefromthecrypt) [inline zipkin-collector/activemq/src/main/java/zipkin2/collector/activemq/ActiveMQCollector.java]: use multi-catch for what is actually thrown, unless literally Exception is thrown

**Review by codefromthecrypt — COMMENTED** [review]
this is good progress. please look at tests for rabbit or kafka to ensure yours are equally comprehensive.

**Review by ghost — APPROVED** [review]
Im ready

**Comment by codefromthecrypt** [comment]
@IAMTJW @moekelley76 can you put your vote on the original issue with thumbsup? https://github.com/openzipkin/zipkin/issues/1990

**Comment by tian-junwei** [comment]
> this is good progress. please look at tests for rabbit or kafka to ensure yours are equally comprehensive.

Is there  a activemq server to be used test ?

**Comment by jbertram** [comment]
How important is performance here, and what's the performance like for this implementation? ActiveMQ 5.x isn't particularly high-performance at this point given its underlying architecture, and storing messages in a database further slows things down. I'm new to Zipkin so I'm curious. Thanks!

**Comment by codefromthecrypt** [comment]
usually performance of the collector is secondary to applicability. for
example, the most applicable transport to a site is a differentiator as it
allows them to use it. a high performance but not applicable answer is just
code to maintain. that is why we do things like "rule of 3" and other
guards to ensure the transports are viable to the most folks.

**Comment by jbertram** [comment]
I was asking about performance more generally and also as it applies to this particular implementation. Is the performance of a collector in general an important factor? Assuming this collector is "applicable," what kind of performance does it provide?

**Comment by codefromthecrypt** [comment]
it is usually the case that storage is slower than the collection
transport. I would be surprised if any activemq option is too slow

**Comment by codefromthecrypt** [comment]
for example some sites are only taking a few hundred messages per second at
100% sample rate.

**Comment by jbertram** [comment]
Is a few hundred messages per second considered near the top-end of the potential volume?

**Comment by codefromthecrypt** [comment]
@jbertram depending on bundling mechanics, higher throughput can be thousands or tens of thousands per second. Higher throughput sites often run multiple collectors though. I would say that most users will not be doing more than a thousand of messages/second.

In the future, this may change as more people explore the costs involved in unsampled (usually this means late sampled, or tiered) collection. For example, we have a https://github.com/adriancole/zipkin-voltdb experiment which can swallow a lot more data than most storage backends. If someone doing 1k messages per second at 1% decides to do 100%, that means multiplying the throughput. For example, some doing what we call "firehose mode" like yelp, they had to write custom collectors as they went over what their kafka setup could do. This is more a thought for you vs what's commonly the case now. I mean to hint there could be some sites doing 100k messages/second into a short term buffer, even if I have no idea when that would be practiced and what the sharding of that rate would typically look like.

Back to reality :)

You'll notice some mentioning the order of thousands of spans per second here. No one currently is using activemq, but that's what this issue is about :) spans should be bundled (ex many spans per message), so if you have say.. 8k spans/second, ideally this is bundled into <1k messages/second of size in orders of low kilobytes.

https://cwiki.apache.org/confluence/display/ZIPKIN/Sites

I'll ask some of the sites to provide more fine grained details on bundling mechanics. Thanks very much for the interest!

**Comment by codefromthecrypt** [comment]
@IAMTJW are you using this in production now (from a fork?) if so, what's your ingest rate?

PS it may be useful to bring this work into a separate repo to evaluate. for example, I made a project like this to help people test/create mongo without affecting the core repository. I can make something like this for you if it would help. https://github.com/adriancole/zipkin-mongodb

**Comment by tian-junwei** [comment]
> @IAMTJW are you using this in production now (from a fork?) if so, what's your ingest rate?
> 
> PS it may be useful to bring this work into a separate repo to evaluate. for example, I made a project like this to help people test/create mongo without affecting the core repository. I can make something like this for you if it would help. https://github.com/adriancole/zipkin-mongodb

We are using this in production(release 2.8.3), but we have only dozens of applications and have not much spans per second here

**Comment by codefromthecrypt** [comment]
@moekelley76 by "I'm ready" does this mean you can try this code? what are you using now and how much throughput of messages do you expect?

**Comment by codefromthecrypt** [comment]
@ImFlog PS we are still waiting for another person to mention they will test this.

As this repository will shortly move to the Apache Software Foundation (ASF) and this is non-trivial code, you should probably prepare by sending a contributor agreement to the ASF. If you have already done this in the past, no work to do. This is a once in life thing.

If you haven't already, easiest way is to download the ICLA template. After filling the [icla.pdf](https://www.apache.org/licenses/icla.pdf) with personal information correctly, print, sign, scan, and send it in mail as an attachment to the secretary (secretary@apache.org)

**Comment by codefromthecrypt** [comment]
whoops I meant @IAMTJW not @ImFlog !

**Comment by tian-junwei** [comment]
> whoops I meant @IAMTJW not @ImFlog !

I  have send mail to secretary@apache.org

**Comment by codefromthecrypt** [comment]
PS this repo is no longer moving to ASF (none are)

**Comment by thanhctES** [comment]
I am waiting for this. Please merge it.

**Comment by codefromthecrypt** [comment]
superceded by https://github.com/openzipkin/zipkin/pull/2639

**Pull request closed without merging**